File: impl/go/hash_join.go
package jed

import (
	"bytes"
	"encoding/binary"
)

// hashJoinTable is a deterministic lookup-only hash table. A hash bucket retains build-row order;
// probe filters the full key bytes so even a forced hash collision cannot admit a false match. The
// executor never iterates the map to emit rows.
type hashJoinTable struct {
	entries     map[uint64][]hashJoinEntry
	hash        func([]byte) uint64
	probeOffset int
}

type hashJoinEntry struct {
	key []byte
	row storedRow
}

func newHashJoinTable(plan *hashJoinPlan, buildOffset, probeOffset int, rows []storedRow, meter *costMeter) (*hashJoinTable, error) {
	return newHashJoinTableWithHash(plan, buildOffset, probeOffset, rows, meter, hashJoinFNV1a)
}

func newHashJoinTableWithHash(plan *hashJoinPlan, buildOffset, probeOffset int, rows []storedRow, meter *costMeter, hasher func([]byte) uint64) (*hashJoinTable, error) {
	t := &hashJoinTable{entries: make(map[uint64][]hashJoinEntry), hash: hasher, probeOffset: probeOffset}
	for _, row := range rows {
		indices := make([]int, len(plan.keys))
		types := make([]dataType, len(plan.keys))
		for i, key := range plan.keys {
			indices[i] = key.right - buildOffset
			types[i] = key.ty
		}
		encoded, present, err := hashJoinRowKey(row, indices, types, costs.HashBuild, meter)
		if err != nil {
			return nil, err
		}
		if !present {
			continue
		}
		h := t.hash(encoded)
		t.entries[h] = append(t.entries[h], hashJoinEntry{key: encoded, row: row})
	}
	return t, nil
}

func (t *hashJoinTable) probe(plan *hashJoinPlan, row storedRow, meter *costMeter) ([]storedRow, error) {
	indices := make([]int, len(plan.keys))
	types := make([]dataType, len(plan.keys))
	for i, key := range plan.keys {
		indices[i] = key.left - t.probeOffset
		types[i] = key.ty
	}
	encoded, present, err := hashJoinRowKey(row, indices, types, costs.HashProbe, meter)
	if err != nil || !present {
		return nil, err
	}
	entries := t.entries[t.hash(encoded)]
	out := make([]storedRow, 0, len(entries))
	for _, entry := range entries {
		if err := meter.Guard(); err != nil {
			return nil, err
		}
		work := len(entry.key)
		if len(encoded) < work {
			work = len(encoded)
		}
		if work == 0 {
			work = 1
		}
		meter.Charge(costs.HashProbe * int64(work))
		if bytes.Equal(entry.key, encoded) {
			out = append(out, entry.row)
		}
	}
	return out, nil
}

func hashJoinRowKey(row storedRow, indices []int, types []dataType, unit int64, meter *costMeter) ([]byte, bool, error) {
	parts := make([][]byte, len(indices))
	present := true
	for i, index := range indices {
		if err := meter.Guard(); err != nil {
			return nil, false, err
		}
		if row[index].Kind == ValNull {
			meter.Charge(unit)
			present = false
			continue
		}
		part, err := encodeTypedKey(types[i], row[index], nil)
		if err != nil {
			return nil, false, err
		}
		charge := int64(len(part))
		if charge == 0 {
			charge = 1
		}
		meter.Charge(unit * charge)
		parts[i] = part
	}
	if !present {
		return nil, false, nil
	}
	var out []byte
	var size [4]byte
	for _, part := range parts {
		binary.BigEndian.PutUint32(size[:], uint32(len(part)))
		out = append(out, size[:]...)
		out = append(out, part...)
	}
	return out, true, nil
}

func hashJoinFNV1a(key []byte) uint64 {
	const offset uint64 = 14695981039346656037
	const prime uint64 = 1099511628211
	h := offset
	for _, b := range key {
		h ^= uint64(b)
		h *= prime
	}
	return h
}

File: impl/rust/src/executor/hash_join.rs
use super::*;

/// Deterministic lookup-only hash table. Buckets retain build-row indices in input order; probes
/// compare the full key bytes, so forced hash collisions cannot admit false matches. The executor
/// never iterates the map to emit rows.
pub(crate) struct HashJoinTable {
    entries: HashMap<u64, Vec<HashJoinEntry>>,
    hash: fn(&[u8]) -> u64,
    probe_offset: usize,
}

struct HashJoinEntry {
    key: Vec<u8>,
    row: usize,
}

impl HashJoinTable {
    pub(crate) fn build(
        plan: &HashJoinPlan,
        build_offset: usize,
        probe_offset: usize,
        rows: &[Row],
        meter: &mut Meter,
    ) -> Result<Self> {
        Self::build_with_hash(
            plan,
            build_offset,
            probe_offset,
            rows,
            meter,
            hash_join_fnv1a,
        )
    }

    fn build_with_hash(
        plan: &HashJoinPlan,
        build_offset: usize,
        probe_offset: usize,
        rows: &[Row],
        meter: &mut Meter,
        hash: fn(&[u8]) -> u64,
    ) -> Result<Self> {
        let mut table = Self {
            entries: HashMap::new(),
            hash,
            probe_offset,
        };
        let indices: Vec<usize> = plan
            .keys
            .iter()
            .map(|key| key.right - build_offset)
            .collect();
        let types: Vec<&Type> = plan.keys.iter().map(|key| &key.ty).collect();
        for (row_index, row) in rows.iter().enumerate() {
            let Some(key) = hash_join_row_key(row, &indices, &types, COSTS.hash_build, meter)?
            else {
                continue;
            };
            let hash = (table.hash)(&key);
            table.entries.entry(hash).or_default().push(HashJoinEntry {
                key,
                row: row_index,
            });
        }
        Ok(table)
    }

    pub(crate) fn probe(
        &self,
        plan: &HashJoinPlan,
        row: &Row,
        meter: &mut Meter,
    ) -> Result<Vec<usize>> {
        let indices: Vec<usize> = plan
            .keys
            .iter()
            .map(|key| key.left - self.probe_offset)
            .collect();
        let types: Vec<&Type> = plan.keys.iter().map(|key| &key.ty).collect();
        let Some(key) = hash_join_row_key(row, &indices, &types, COSTS.hash_probe, meter)? else {
            return Ok(Vec::new());
        };
        let mut rows = Vec::new();
        if let Some(entries) = self.entries.get(&(self.hash)(&key)) {
            for entry in entries {
                meter.guard()?;
                let work = entry.key.len().min(key.len()).max(1);
                meter.charge(COSTS.hash_probe * i64::try_from(work).unwrap_or(i64::MAX));
                if entry.key == key {
                    rows.push(entry.row);
                }
            }
        }
        Ok(rows)
    }
}

fn hash_join_row_key(
    row: &Row,
    indices: &[usize],
    types: &[&Type],
    unit: i64,
    meter: &mut Meter,
) -> Result<Option<Vec<u8>>> {
    let mut parts = Vec::with_capacity(indices.len());
    let mut present = true;
    for (&index, ty) in indices.iter().zip(types) {
        meter.guard()?;
        if matches!(row[index], Value::Null) {
            meter.charge(unit);
            present = false;
            parts.push(Vec::new());
            continue;
        }
        let part = encode_typed_key(ty, &row[index], None)?;
        meter.charge(unit * i64::try_from(part.len().max(1)).unwrap_or(i64::MAX));
        parts.push(part);
    }
    if !present {
        return Ok(None);
    }
    let mut out = Vec::new();
    for part in parts {
        let len = u32::try_from(part.len()).expect("a row value fits the u32 key length contract");
        out.extend_from_slice(&len.to_be_bytes());
        out.extend_from_slice(&part);
    }
    Ok(Some(out))
}

fn hash_join_fnv1a(key: &[u8]) -> u64 {
    let mut hash = 14_695_981_039_346_656_037_u64;
    for byte in key {
        hash ^= u64::from(*byte);
        hash = hash.wrapping_mul(1_099_511_628_211);
    }
    hash
}

#[cfg(test)]
mod tests {
    use super::*;

    fn collide(_: &[u8]) -> u64 {
        0
    }

    #[test]
    fn forced_hash_collisions_recheck_full_keys() {
        let plan = HashJoinPlan {
            keys: vec![HashJoinKey {
                left: 0,
                right: 1,
                ty: Type::Scalar(ScalarType::Int32),
            }],
        };
        let rows = vec![
            vec![Value::Int(1), Value::Int(10)],
            vec![Value::Int(2), Value::Int(20)],
            vec![Value::Int(2), Value::Int(21)],
        ];
        let mut meter = Meter::new();
        let table =
            HashJoinTable::build_with_hash(&plan, 1, 0, &rows, &mut meter, collide).unwrap();
        let got = table
            .probe(&plan, &vec![Value::Int(2)], &mut Meter::new())
            .unwrap();
        assert_eq!(got, vec![1, 2]);
    }
}
